split amazon title into title + subtitle and remove parens

hornc opened the pull request:
closes #2311

### Description
<!-- What does this PR achieve? [feature|hotfix|fix|refactor] -->

Closes #

### Technical
<!-- What should be noted about the implementation? -->

### Testing
<!-- Steps for reviewer to reproduce/verify what this PR does/fixes. -->

### Evidence
<!-- If this PR touches UI, please post evidence (screenshots) of it behaving correctly. -->

hornc commented on openlibrary/core/vendors.py:
```diff
@@ -187,6 +206,16 @@ def clean_amazon_metadata_for_load(metadata):
+        conforming_metadata['subtitle'] = subtitle
+    # Record original title if some content has been removed (i.e. parentheses)
+    if metadata['title'] != conforming_metadata.get('full_title', conforming_metadata['title']):
+        conforming_metadata['notes'] = "Source title: %s" % metadata['title']
```
perhaps only set this if `( )` were present in the original

cdrini replied:
I think I like your current `if`-check better; checking for `(` makes this code rely on the internals of `split_amazon_title`. Just checking if it's different has better decoupling (and is easier to read).

hornc replied:
ok, you have convinced me :)

cdrini commented on openlibrary/core/vendors.py:
```diff
@@ -164,6 +164,25 @@ def _get_amazon_metadata(id_=None, id_type='isbn'):
+    """Splits an Amazon title into (title, subtitle),
+    strips parenthetical tags.
+    :param str full_title:
+    :rtype: tuple
```
```suggestion
    :rtype: (str, str | None)
```

cdrini commented on openlibrary/core/vendors.py:
```diff
@@ -164,6 +164,25 @@ def _get_amazon_metadata(id_=None, id_type='isbn'):
+    :return: (title, subtitle | None)
+    """
+    # strip parens
+    full_title = re.sub('\([^\)]*\)', '', full_title)
```
This regex won't work for nested parens

cdrini commented on openlibrary/core/vendors.py:
```diff
@@ -164,6 +164,25 @@ def _get_amazon_metadata(id_=None, id_type='isbn'):
+    """
+    # strip parens
+    full_title = re.sub('\([^\)]*\)', '', full_title)
+    re_split_title = re.compile(
```
It looks like this regex is _also_ removing parents; maybe let's flip this around? That would entirely remove this complex regex. So first split into title/subtitle, and then remove the parens on each. (`title_parts = full_title.split(': ', 1)`, then `trim`/parens regex)

hornc replied:
removing the complex regex :+1: I copied it from existing OL code, and while it looks like it is doing something with parenthesis, it doesn't seem to affect any examples I have found, so simplifying and making what it needs to do clearer is a good suggestion!

cdrini commented on openlibrary/core/vendors.py:
```diff
@@ -187,6 +206,16 @@ def clean_amazon_metadata_for_load(metadata):
         asin = metadata.get('source_records')[0].replace('amazon:', '')
         conforming_metadata['identifiers'] = {'amazon': [asin]}
+    title, subtitle = split_amazon_title(metadata['title'])
+    conforming_metadata['title'] == title
```
```suggestion
    conforming_metadata['title'] = title
```

hornc replied:
oh wow, this could be the cause of the weird results I saw in some cases, great catch!

cdrini commented on openlibrary/core/vendors.py:
```diff
@@ -187,6 +206,16 @@ def clean_amazon_metadata_for_load(metadata):
+    title, subtitle = split_amazon_title(metadata['title'])
+    conforming_metadata['title'] == title
+    if subtitle:
+        conforming_metadata['full_title'] = title + ' : ' + subtitle
```
I think this should be set to `title` outside of this `if` statement; that way `full_title` is always defined.

hornc replied:
I was avoiding setting `full_title` unless it was different from `title`, which seems to be the current way of doing things. I might leave this until we have a clearer picture of how `full_title` is used.

cdrini commented on openlibrary/core/vendors.py:
```diff
@@ -187,6 +206,16 @@ def clean_amazon_metadata_for_load(metadata):
+    title, subtitle = split_amazon_title(metadata['title'])
+    conforming_metadata['title'] == title
+    if subtitle:
+        conforming_metadata['full_title'] = title + ' : ' + subtitle
```
```suggestion
        conforming_metadata['full_title'] = title + ': ' + subtitle
```

hornc replied:
This is a funny one as I added the space deliberately to match what I thought was being done in the MARC import case, I'd need to confirm it, but I thought there was cataloguing practice that did things this way. I does have implications for detecting whether the title was modified in one of the comment areas further up, and I want to avoid saving title : subtitle in the notes field if there were no other changes. I'll check to see if there is a compelling reason to drop the extra space, and make sure we are consistent with it.

cdrini replied:
I don't feel strongly about this; if the user never sees this text, feel free to resolve this as "Won't fix" :)

hornc replied:
I _was_ thinking about a real thing -- ISBD cataloguing practice and punctuation https://www.ifla.org/files/assets/cataloguing/isbd/isbd-cons_20110321.pdf which is / influences the punctuation used in MARC records. I've set a new constant `ISBD_UNIT_PUNCT` to ` : ` to make it look less like a typo, and have used it in a couple of places :)

cdrini commented on openlibrary/core/vendors.py:
```diff
@@ -187,6 +206,16 @@ def clean_amazon_metadata_for_load(metadata):
+    conforming_metadata['title'] == title
+    if subtitle:
+        conforming_metadata['full_title'] = title + ' : ' + subtitle
+        conforming_metadata['title'] = title
```
This is already done, no?
```suggestion
```

cdrini commented on openlibrary/tests/core/test_vendors.py:
```diff
@@ -41,16 +40,34 @@ def test_clean_amazon_metadata_for_load_ISBN():
     assert result.get('offer_summary') is None
 
 
+amazon_titles = [
```
Yep; note even if you're not sure if it's the "correct" thing, having a test to document what it's doing is useful, because other code might rely on this function to behave in a certain way.

hornc replied:
Thanks for the reviews @cdrini and @tfmorris , I have dropped the title splitting regex in favour of @cdrini's suggestion of using split, and crafted a regex to strip parentheses where they occur that can handle one level of nesting (like in that example) cleanly.  It's a bit complex, but the tests cover what it can do.

hornc commented:
@cdrini I think I have addressed all the comments. I've switched the regexes around, simplified one part, and made another more complex... hopefully a net positive change!
